### PR TITLE
90%: Metatron support

### PR DESCRIPTION
**Pull request by rsinger**
Provides a DSL to metatron_rb's interface to `GET /2/works` and abstracts the swagger/JSON-API models away.

How to use:

```ruby

> works = Talis::Bibliography::Work.find(query: 'some query', offset: 0, limit: 10, include: ['manifestations.assets'])

=> #<MetatronClient::WorkResultSet:0x007fb8841687a0 @meta=#<MetatronClient::ResultSetMeta:0x007fb884168480 @offset=0.0, @limit=10.0, @count=2.0>, @data=[#<Talis::Bibliography::Work:0x007fb884360b98 @work_data=#<MetatronClient::WorkData:0x007fb88416b748 @id="anatomy_of_a_frog/authora", @type="works", @attributes=#<MetatronClient::WorkAttributes:0x007fb884363c30 @title="Anatomy of a Frog">, @relationships=#<MetatronClient::WorkRelationships:0x007fb884363988 @manifestations=#<MetatronClient::OneToManyRelationship:0x007fb884363780 @data=[#<MetatronClient::ResourceLink:0x007fb8843634b0 @id="abc:0001", @type="manifestations">, #<MetatronClient::ResourceLink:0x007fb8843631e0 @id="abc:0002", @type="manifestations">]>>>, @title="Anatomy of a Frog", @manifestations=[#<MetatronClient::ManifestationData:0x007fb884360198 @id="abc:0001", @type="manifestations", @attributes={:identifiers=>{:isbn=>["9780123456789"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"1998"}, :title=>"Anatomy of a Frog"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}, :assets=>{:data=>[{:id=>"xyz:0002", :type=>"assets"}]}}>, #<MetatronClient::ManifestationData:0x007fb8871c7748 @id="abc:0002", @type="manifestations", @attributes={:identifiers=>{:isbn=>["978012345678X"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"2005"}, :title=>"Anatomy of a Frog", :edition_statment=>"2nd edition"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}}>], @assets=[#<MetatronClient::AssetData:0x007fb8871c7b80 @id="xyz:0002", @type="assets", @attributes={:property=>"value of 0002"}>]>, #<Talis::Bibliography::Work:0x007fb8871c7478 @work_data=#<MetatronClient::WorkData:0x007fb884362f10 @id="froggy_went_a_courtin/bloggsj", @type="works", @attributes=#<MetatronClient::WorkAttributes:0x007fb884362b78 @title="Froggy Went A Courtin'">, @relationships=#<MetatronClient::WorkRelationships:0x007fb8843628d0 @manifestations=#<MetatronClient::OneToManyRelationship:0x007fb8843626c8 @data=[#<MetatronClient::ResourceLink:0x007fb8843623f8 @id="abc:0003", @type="manifestations">]>>>, @title="Froggy Went A Courtin'", @manifestations=[#<MetatronClient::ManifestationData:0x007fb8871c68c0 @id="abc:0003", @type="manifestations", @attributes={:identifiers=>{:isbn=>["9789876543210"]}, :place_of_publication=>"Tennessee", :publishing_status=>"00", :publisher=>{:imprint=>"Song books", :name=>"Song books"}, :country=>"xxu", :dates=>{:published=>"1988"}, :title=>"Froggy Went a Courtin'"}, @relationships={:works=>{:data=>[{:id=>"froggy_went_a_courtin/bloggsj", :type=>"works"}]}}>]>], @included=[#<MetatronClient::ResourceData:0x007fb884362060 @id="abc:0001", @type="manifestations", @attributes={:identifiers=>{:isbn=>["9780123456789"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"1998"}, :title=>"Anatomy of a Frog"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}, :assets=>{:data=>[{:id=>"xyz:0002", :type=>"assets"}]}}>, #<MetatronClient::ResourceData:0x007fb884361c50 @id="abc:0002", @type="manifestations", @attributes={:identifiers=>{:isbn=>["978012345678X"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"2005"}, :title=>"Anatomy of a Frog", :edition_statment=>"2nd edition"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}}>, #<MetatronClient::ResourceData:0x007fb884361840 @id="abc:0003", @type="manifestations", @attributes={:identifiers=>{:isbn=>["9789876543210"]}, :place_of_publication=>"Tennessee", :publishing_status=>"00", :publisher=>{:imprint=>"Song books", :name=>"Song books"}, :country=>"xxu", :dates=>{:published=>"1988"}, :title=>"Froggy Went a Courtin'"}, @relationships={:works=>{:data=>[{:id=>"froggy_went_a_courtin/bloggsj", :type=>"works"}]}}>, #<MetatronClient::ResourceData:0x007fb884361430 @id="xyz:0001", @type="assets", @attributes={:property=>"value of 0001"}, @relationships={:manifestations=>{:data=>[{:id=>"abc:0003", :type=>"manifestations"}]}}>, #<MetatronClient::ResourceData:0x007fb884361020 @id="xyz:0002", @type="assets", @attributes={:property=>"value of 0002"}, @relationships={:manifestations=>{:data=>[{:id=>"abc:0001", :type=>"manifestations"}]}}>]>

> works.first.id
=> "anatomy_of_a_frog/authora"

> works.first.title
=> "Anatomy of a Frog"

> works.first.manifestations
=> [#<MetatronClient::ManifestationData:0x007fb884360198 @id="abc:0001", @type="manifestations", @attributes={:identifiers=>{:isbn=>["9780123456789"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"1998"}, :title=>"Anatomy of a Frog"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}, :assets=>{:data=>[{:id=>"xyz:0002", :type=>"assets"}]}}>, #<MetatronClient::ManifestationData:0x007fb8871c7748 @id="abc:0002", @type="manifestations", @attributes={:identifiers=>{:isbn=>["978012345678X"]}, :place_of_publication=>"Birmingham", :publishing_status=>"00", :publisher=>{:imprint=>"Talis Educational Press", :name=>"Talis Information Limited"}, :country=>"xku", :dates=>{:published=>"2005"}, :title=>"Anatomy of a Frog", :edition_statment=>"2nd edition"}, @relationships={:works=>{:data=>[{:id=>"anatomy_of_a_frog/authora", :type=>"works"}]}}>]

> works.first.assets
=> [#<MetatronClient::AssetData:0x007fb8871c7b80 @id="xyz:0002", @type="assets", @attributes={:property=>"value of 0002"}>]

```

This only 're-models' works: manifestations and assets should be addressed when we tackle those routes directly in the API.  It replaces the MetatronClient::WorkData object with a Talis::Bibliography::Work object which organizes its data a little more intuitively.

When we integrate Atlas' PaginatedJsonApiArray object into Talis_rb, we should replace `Talis::Bibliography::ResultSet` and `MetatronClient::WorkResultSet` with it.  In the meantime, we should add the functionality we want from PaginatedJsonApiArray to ResultSet.

TODO:
- [x] Specs
- [x] Point at a stable Metatron with talis/platform#333 and talis/platform#332 implemented
- [x] Release metatron_rb and update talis.gemspec/Gemfile

Closes talis/platform#334

### CODE REVIEW

**Comment by rsinger** [comment]
@pads Theoretically, all this needs to get a green build is to point it at a legitimately working Metatron endpoint that supports oauth and includes.  Good luck.

**Comment by pads** [comment]
Happy holidays! 🍹 🍹 🍹

**Comment by pads** [comment]
@rsinger this is good to merge now - will leave it for a bit if you want to give this a once-over?

File: .travis.yml [hunk 1]
```diff
@@ -14,5 +14,8 @@ env:
   global:
   - PERSONA_TEST_HOST: https://staging-users.talis.com
   - BLUEPRINT_TEST_HOST: https://staging-hierarchy.talis.com
+  - METATRON_TEST_HOST: https://4yk6o0puk0.execute-api.eu-west-1.amazonaws.com
+  - METATRON_BASE_PATH: /development
```
Comment (rsinger) [inline .travis.yml]: How will this get reconciled when we have a production instance?  What would this be set to?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    bundle install
+    
+Create a `.env` file and add the following:
+
+    PERSONA_TEST_HOST=<your local persona host>
+    PERSONA_OAUTH_CLIENT=<your local persona client ID (with su scope)>
+    PERSONA_OAUTH_SECRET=<your local persona client secret (with su scope)>
+    BLUEPRINT_TEST_HOST=<your local blueprint host>
+    METATRON_TEST_HOST=<your local metatron host>
+    METATRON_BASE_PATH=<if not a production env, use /development>
```
Comment (rsinger) [inline CONTRIBUTING.md]: This begins to explain it, I guess -- so it would be `/2` (or whatever) in production?

File: Gemfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
 source 'https://rubygems.org'
 
-# Specify your gem's dependencies in talis.gemspec
+# Specify your gem's dependencies in metatron_ruby_client.gemspec
```
Comment (rsinger) [inline Gemfile]: Copy and paste error, I think -- this file looks like it shouldn't have been changed.

File: lib/talis/bibliography/result_set.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      end
+
+      private
+
+      def model_data
+        data.each_with_index do |data, i|
+          resource = Work.new data if is_a? MetatronClient::WorkResultSet
+
+          resource.hydrate_relationships(included) if included
+          self.data[i] = resource
```
Comment (rsinger) [inline lib/talis/bibliography/result_set.rb]: Is `self.` necessary?

File: lib/talis/bibliography/work.rb [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+      # rubocop:disable Metrics/LineLength
+      class << self
+        # Search for bibliographic works
+        # @param request_id [String] ('uuid') unique ID for the remote request.
+        # @param query [String] the query to filter works on
+        # @param offset [Integer] the page offset
+        # @param limit [Integer] the number of works per page
+        # @param include [Array] the related resources to associate with each work
+        #   see {https://github.com/talis/blueprint_rb/blob/master/docs/HierarchyApi.md#search_nodes}
```
Comment (rsinger) [inline lib/talis/bibliography/work.rb]: Copy and paste error - this should point to metatron_rb's docs.

File: lib/talis/bibliography/work.rb [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+          rescue Talis::Errors::NotFoundError
+            meta = OpenStruct.new(offset: offset, limit: limit, count: 0)
+            MetatronClient::WorkResultSet.new(data: [], meta: meta)
+          end
+        end
+
+        # Fetch a single work by id
+        # @param request_id [String] ('uuid') unique ID for the remote request.
+        # @param id [String] the ID of the work to fetch.
+        # @param type [String] the type of node to fetch.
```
Comment (rsinger) [inline lib/talis/bibliography/work.rb]: Copy and paste error - these aren't part of this method.

File: .travis.yml (hunk 1 above)
Comment (pads) [inline .travis.yml]: It's an optional env var - there is no base path in production.

File: CONTRIBUTING.md (hunk 2 above)
Comment (pads) [inline CONTRIBUTING.md]: There is no base path in production - the `/2` is included in all URLs as a result of the swagger code gen.

File: lib/talis/bibliography/result_set.rb (hunk 4 above)
Comment (pads) [inline lib/talis/bibliography/result_set.rb]: No but Swagger code gen thought so. I'm not gonna mess with this level of code generation!

File: CONTRIBUTING.md (hunk 2 above)
Comment (pads) [inline CONTRIBUTING.md]: See https://github.com/talis/metatron_rb/blob/master/stub-service/swagger.yaml and https://github.com/talis/metatron_rb/blob/master/stub-service/api/default_api.rb#L4

I used the swagger doc generated by API gateway and this is how it works. The downside of this is that the code-gen generates method names with a `2_` prefix which is a syntax error in Ruby - so I had to find/replace those by hand.

**Comment by rsinger** [comment]
:shipit:

**Pull request merged**